app/src/main/java/com/example/habitsmasher/UserRegistrationFragment.java

```diff
@@ -0,0 +1,213 @@
+import android.widget.ProgressBar;
+import android.widget.Toast;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.appcompat.app.ActionBar;
+import androidx.appcompat.app.AppCompatActivity;
+import androidx.fragment.app.Fragment;
+import androidx.navigation.NavController;
+import androidx.navigation.fragment.NavHostFragment;
+
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+import java.util.HashMap;
+
+/**
+ * This class holds the front-end elements related to the user sign up page
+ * Author: Rudy Patel
+ */
+public class UserRegistrationFragment extends Fragment {
+    private static final String USER_REGISTERED_MESSAGE = "User registered!";
+    private static final String FAILED_TO_ADD_USER_MESSAGE = "Failed to add user, try again!";
+    private static final String FAILED_TO_REGISTER_MESSAGE = "Failed to register";
+
+    private FirebaseAuth _auth;
+    private ProgressBar _progressBar;
+
+    @Nullable
+    @Override
+    public View onCreateView(@NonNull LayoutInflater inflater,
+                             @Nullable ViewGroup container,
+                             @Nullable Bundle savedInstanceState) {
+
+        View view = inflater.inflate(R.layout.user_registration, container, false);
+
+        _auth = FirebaseAuth.getInstance();
+
+        // get front-end elements
+        EditText emailInput = view.findViewById(R.id.registration_email);
+        EditText passwordInput = view.findViewById(R.id.registration_password);
+        EditText usernameInput = view.findViewById(R.id.registration_username);
+        Button registerButton = view.findViewById(R.id.registration_signup_button);
+        Button backToLoginButton = view.findViewById(R.id.registration_go_back_to_login_button);
+        _progressBar = view.findViewById(R.id.registration_progress_bar);
+
+        setClickListenerForBackToLoginButton(backToLoginButton);
+
+        setClickListenerForRegisterButton(emailInput, passwordInput, usernameInput, registerButton);
+
+        return view;
+    }
+
+    /**
+     * This method sets the click listener for the register button, along with routing to the
+     * associated validators to process the input data
+     * @param emailInput the email input box
+     * @param passwordInput the password input box
+     * @param usernameInput the username input box
+     * @param registerButton the register button
+     */
+    private void setClickListenerForRegisterButton(EditText emailInput,
+                                                   EditText passwordInput,
+                                                   EditText usernameInput,
+                                                   Button registerButton) {
+        registerButton.setOnClickListener(new View.OnClickListener() {
```

Comment:
Again, why not initialize the button inside of the function and not pass it as a parameter?

Reply:
I think we can just pick **one** way and stick with it. Right now, I'm just choosing to **initialize** beforehand and pass the reference to the button to set the listener.

This should be fine.

The other way you're suggesting is equally correct, but it's not the biggest deal, to be honest, and I don't want to lose the current approvals.